LanguageFeatures/Patterns/type_inference_A32_t02.dart

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// To type check a pattern p being matched against a value of type M:
+/// ...
+/// If p with required type T is in an irrefutable context:
+/// - It is a compile-time error if M is not assignable to T.
+/// - Else if M is not a subtype of T then an implicit coercion or cast is
+///   inserted before the pattern binds the value, tests the value's type,
+///   destructures the value, or invokes a function with the value as a target
+///   or argument.
+///
+/// @description Check that it is a compile-time error if `M` is not assignable
+/// to `T` in irrefutable context. Test list patterns
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns
+
+main() {
+  var <int>[x1, num x2] = <num>[1, 2];
+//                        ^^^^^^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+
+  final [x3, num x4] = ["42", true];
```

Comment:
We need to use the simplest and most firm way to obtain a specific required type. I believe this would be achieved by using explicit type arguments on the pattern. So please use explicit types on the patterns everywhere in this test.

For a list and a map it's the type arguments at the top level, for a record it's specified for each field, so that could just be a declared type of a variable pattern. We could have `var <num>[int i] = ...;`, but that's a different error (it's an error for a variable pattern to have a declared type which is not statically known to be a supertype of the matched object type). So we don't need to test that here, either.

We have other tests where it is tested how the required type of the pattern is computed when it is not simply specified as explicit type arguments, and there's no need to duplicate this kind of testing here.

For the initializing expression I actually think it makes sense to write all types explicitly as well. The point is that (presumably) there is no way to obtain a different set of matched value types at all the levels that the pattern explores, other than those which can be expressed using explicit types: `["42", true]` will get one or the other type argument based on the pattern type schema and/or the elements, and we can always write an explicit type argument which is the same as that which would arise because of some complex interaction between elements and the pattern schema. No need to do that, we just need to force the type to be something specific, in the most firm way we can find.

This means that there are fewer cases to test here, but do consider `var <S>[...] = <T>[...]` with all relevant relationships between `S` and `T`: Same type, subtype, supertype, `dynamic`. I think this test does not depend on the elements in the initializing expression: They can be wrong (but that's a different error), and they can be correct (and that's what we want), but they can't change the matched value type anywhere in the pattern. So no tricks are needed in order to make those elements explicitly typed.